go/merge/three_way.go

```diff
@@ -200,7 +190,81 @@ func threeWayStructMerge(a, b, parent types.Struct, vwr types.ValueReadWriter, p
 		}
 		panic(fmt.Errorf("Bad key type in diff: %s", change.V.Type().Describe()))
 	}
-	return threeWayOrderedSequenceMerge(parent, aDiff, bDiff, makeGetFunc(a), makeGetFunc(b), makeGetFunc(parent), apply, vwr, progress)
+	return m.threeWayOrderedSequenceMerge(structCandidate{a}, structCandidate{b}, structCandidate{parent}, apply, path)
+}
+
+type candidate interface {
```

Comment:
My gut reaction is that that will result is lots of type-switching, because `a`, `b`, and `parent` will come in as a types.Value. I feel like that's lame, because type-appropriate implementations of the same method is like...literally the thing you're supposed to use interfaces for.

What is it that's distracting? Is it that the definitions are all kind of in-line here? Is it the apply functions having to sorta un-candidate them? In my mind, the stuff that's complicated is all over in `threeWayOrderedSequenceMerge()`, and this use of adaptors doesn't change anything over there. This is really just shoehorning three different things into a similar enough shape that we can run them all through a single, common, complex code path.